### PR TITLE
Tighten up branch and reinit; define ResumptionPSKUsage

### PR DESCRIPTION
**Pull request by bifurcation**
This PR reflects further discussion on #502 and replaces #518.  The high-level impacts are:

* Add clearer, more prescriptive definitions of the re-init and branch operations
* Use a single syntax for all resumption PSKs, but also have a `usage` parameter to distinguish cases
* Allow an `application` usage to cover any other cases defined by applications

Once #520 lands, we should also add some text to the protocol overview to describe these operations.  Pasting some text here for usage in that future PR...

```
## Relationships Between Epochs

A group comprises a single linear sequence of epochs. However, it can sometimes
be useful to express relationships between epochs, either within a group or
across groups.  MLS derives a resumption pre-shared key (PSK) from each epoch to
allow a entropy extracted from one epoch to be injected into a future epoch,
effectively creating a causal link between the two epochs.

MLS supports two ways to tie a new group to an existing group. Re-initialization
closes one group and creates a new group comprising the same members with
different parameters. Branching starts a new group with a subset of the original
group's participants (with no effect on the original group).  In both cases,
the new group is linked to the old group via a resumption PSK.

~~~~~
epoch_A_[n-1]
     |
     |
     |<-- ReInit
     |
     V
epoch_A_[n]           epoch_B_[0]
     .                     |
     .  PSK(usage=reinit)  |
     .....................>|
                           |
                           V
                      epoch_B_[1]
~~~~~
{: title="Reinitializing a group" }


~~~~~
epoch_A_[n-1]
     |
     |
     |<-- ReInit
     |
     V
epoch_A_[n]           epoch_B_[0]
     |                     |
     |  PSK(usage=branch)  |
     |....................>|
     |                     |
     V                     V
epoch_A_[n+1]         epoch_B_[1]
~~~~~
{: title="Branching a group" }

Applications may also choose to use resumption PSKs to link epochs in other
ways.  For example, the following figure shows a case where a resumption PSK
from epoch `n` is injected into epoch `n+k`.  This demonstrates that the members
of the group at epoch `n+k` were also members at epoch `n` (irrespective of any
changes to these members' keys due to Updates or Commits.

~~~~~
epoch_A_[n-1]
     |
     |
     |<-- ReInit
     |
     V
epoch_A_[n]
     |                     
     |  PSK(usage=application)  
     |.....................
     |                    . 
     |                    . 
    ...                  ...
     |                    . 
     |                    . 
     V                    .
epoch_A_[n+k-1]           .
     |                    . 
     |                    . 
     |<.................... 
     |                      
     V                     
epoch_A_[n+k]       
~~~~~
{: title="Reinjecting entropy from an earlier epoch" }
```

### CODE REVIEW

File: draft-ietf-mls-protocol.md [hunk 1]
```diff
@@ -2323,25 +2314,52 @@ extensions can be updated, a GroupContextExtensions proposal is invalid if it
...
+in the Welcome message that creates the branched subgroup.
+
+A client receiving a Welcome including a PreSharedKey of type `resumption` with
+usage `branch` MUST verify that the new group reflects a subgroup branched from
+the referenced group.
+
+* The `version` and `ciphersuite` values in the Welcome MUST be the same as
+  those used by the old group.
+* Each KeyPackage in a leaf node of the new group's tree MUST be a leaf in the
+  old group's tree at the epoch indicated in the PreSharedKey.
```
Comment (kkohbrok) [inline draft-ietf-mls-protocol.md:2355]: Do we really want the KeyPackages to be identical? This seems to constitute key-reuse across groups. This is where a (identity,endpoint_id) tuple would come in handy ;-) But I think going with the new multi-identity credentials, we could mandate here that the application has to ensure that the participants in the new group are a subset of the participants in the old group

**Review by kkohbrok — COMMENTED** [review]
Looks good! Although I'm not sure in the branching keys we want to have the exact same key packages as in the main group.

**Comment by kkohbrok** [comment]
One question regarding the diagram in the PR description. In the re-init case, shouldn't the "B" group start at epoch `n+1`? At least that's what the PR specifies in the spec.

**Comment by TWal** [comment]
I think the re-init diagram is fine because the spec says

    The `epoch` in the Welcome message MUST be 1

However in the branch case (and application case) diagram I don't think there should be a ReInit proposal? (At least it is not said in the spec).

Also, it is not entirely clear how to construct a `PreSharedKeyID` in the branch case (or application case): what values should have `psk_group_id` and `psk_epoch`?

**Comment by kkohbrok** [comment]
Ah, sorry I misread. You are right and the re-init diagram is fine the way it is.

**Review by raphaelrobert — COMMENTED** [review]
I think this looks fine generally (just one comment)

**Comment by br-hale** [comment]
Minor typo in the text above: "allow a entropy" -> "allow entropy".

**Pull request merged**